tutorials/composite_bo_with_hogp.ipynb

```diff
@@ -0,0 +1,646 @@
+    "    # X = [M, D, L, tau]\n",
+    "    bounds = torch.tensor(\n",
+    "        [[7.0, 0.02, 0.01, 30.010], [13.0, 0.12, 3.00, 30.295]],\n",
+    "        device=device,\n",
+    "        dtype=dtype,\n",
+    "    )\n",
+    "\n",
+    "    M0 = torch.tensor(10.0, device=device, dtype=dtype)\n",
+    "    D0 = torch.tensor(0.07, device=device, dtype=dtype)\n",
+    "    L0 = torch.tensor(1.505, device=device, dtype=dtype)\n",
+    "    tau0 = torch.tensor(30.1525, device=device, dtype=dtype)\n",
+    "\n",
+    "    # we can vectorize everything, no need for loops\n",
+    "    if s_size == 3:\n",
+    "        S = torch.tensor([0.0, 1.0, 2.5], device=device, dtype=dtype)\n",
+    "    else:\n",
+    "        S = torch.linspace(0.0, 2.5, s_size, device=device, dtype=dtype)\n",
+    "    if t_size == 4:\n",
+    "        T = torch.tensor([15.0, 30.0, 45.0, 60.0], device=device, dtype=dtype)\n",
+    "    else:\n",
+    "        T = torch.linspace(15.0, 60.0, t_size, device=device, dtype=dtype)\n",
+    "\n",
+    "    Sgrid, Tgrid = torch.meshgrid(S, T)\n",
+    "\n",
+    "    # X = [M, D, L, tau]\n",
+    "    def c_batched(X, k=None):\n",
+    "        return torch.stack([env_cfun(Sgrid, Tgrid, *x) for x in X])\n",
+    "\n",
+    "    c_true = env_cfun(Sgrid, Tgrid, M0, D0, L0, tau0)\n",
+    "\n",
+    "    def neq_sum_quared_diff(samples):\n",
+    "        # unsqueeze\n",
+    "        if samples.shape[-1] == (s_size * t_size):\n",
+    "            samples = samples.unsqueeze(-1).reshape(\n",
+    "                *samples.shape[:-1], s_size, t_size\n",
+    "            )\n",
+    "\n",
+    "        sq_diffs = (samples - c_true).pow(2)\n",
+    "        return sq_diffs.sum(dim=(-1, -2)).mul(-1.0)\n",
+    "\n",
+    "    objective = GenericMCObjective(neq_sum_quared_diff)\n",
+    "    num_samples = 32\n",
+    "\n",
+    "    return c_batched, objective, bounds, num_samples"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## BO Loop"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, we run the BO loop over three trials for $15$ batches each. This loop might take a while."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_init = 30\n",
+    "n_batches = 15\n",
+    "batch_size = 4\n",
+    "n_trials = 3"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from botorch.optim.fit import fit_gpytorch_torch"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As a word of caution, we've found that when fitting the HOGP model, using first-order optimizers (e.g. Adam) as is used in `fit_gpytorch_torch` tends to outperform second-order optimizers such as L-BFGS-B due to the large number of free parameters in the HOGP."
```

Comment:
Is the reason that L-BFGS-B fails or that it is "too good" and tends to overfit a model that might not be sufficiently regularized?

Reply:
That's what it was looking like.

Reply:
I see. Maybe worth mentioning sth along those lines so his doesn't seem like magic. Adding proper regularizations sounds like a "future work" item...